lnwallet/channel.go

```diff
@@ -1704,6 +1705,134 @@ func (lc *LightningChannel) SignNextCommitment() ([]byte, error) {
 	return sig, nil
 }
 
+// ReceiveReestablish is used to handle the remote channel reestablish message
+// and generate the set of updates which are have to be sent to remote side
+// to synchronize the states of the channels.
+func (lc *LightningChannel) ReceiveReestablish(msg *lnwire.ChannelReestablish) (
+	[]lnwire.Message, error) {
+
+	lc.Lock()
+	defer lc.Unlock()
+
+	// As far we store on last commitment transaction we should rely on the
+	// height of the commitment transaction in order to calculate the length.
+	numberRemoteCommitments := lc.remoteCommitChain.tip().height + 1
+
+	// Number of the revocations might be calculated as the height of the
+	// commitment transactions which will be revoked next minus one. And plus
+	// one because height starts from zero.
+	numberRemoteRevocations := lc.remoteCommitChain.tail().height - 1 + 1
+
+	var updates []lnwire.Message
+	commitmentChainDiff := msg.NextLocalCommitmentNumber - numberRemoteCommitments
+	if commitmentChainDiff == 0 {
+		// If remote side expects as receive commitment which we already
+		// consider as last, than it means that they aren't received our
+		// last commit sig message.
+		commitment := lc.remoteCommitChain.tip()
+		chanID := lnwire.NewChanIDFromOutPoint(lc.channelState.ChanID)
+		for _, htlc := range commitment.outgoingHTLCs {
```

Comment:
My work re-writing the portions of the state machine also caused be to realize another flaw in our current persistence: we don't store the _exact_ state of the remote party's commitment chain. This can be remedied by syncing the exact state of the remote part's commitment when we receive a `RevokeAndAck` message. Currently, we only store the revocation state of the now discarded commitment transaction. But we also need to store the state of their _new_ commitment state in order to properly restart.